### PR TITLE
[SW-2189] Generate Algorithm Wrappers in Scala From H2O Code

### PR DESCRIPTION
**Pull request by mn-mikke**

### CODE REVIEW

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterResolver.scala [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ai.h2o.sparkling.api.generation.common
+
+trait ParameterResolver {
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterResolver.scala:20]: Note: There will be some exceptions in namings, for example in GLM the param name is alpha but in SW we have alphaValue. But there should be really small number of exceptions like this.

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+         |      Map(
+         |${generateH2OAssignments(parameters)})
+         |  }
+      """.stripMargin
+    }
+  }
+
+  private def generateParameterDefinitions(parameters: Seq[Parameter]): String = {
+    parameters.map { parameter =>
+      s"""  private val ${parameter.name} = ${resolveParameterType(parameter)}Param("${parameter.name}")"""
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala:73]: Maybe we can also put here doc from H2O

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/Parameter.scala [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ai.h2o.sparkling.api.generation.common
+
+case class Parameter(
+  name: String,
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/Parameter.scala:21]: Do I understand correctly that `name` is SW name and `originalName` is H2O name?

Great idea! Right now we don't store the mapping in the code but could be used in few places -> mainly Grid for the SW -> h2o param mapping

Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/Parameter.scala:21]: I will rename it to be more descriptive, thanks!

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterResolver.scala (hunk 1 above)
Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterResolver.scala:20]: Thanks for heads up!

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala (hunk 2 above)
Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala:73]: ok, will do

**Comment by jakubhava** [comment]
just fyi: The infra should be stable now, failing on linter

**Comment by mn-mikke** [comment]
@jakubhava The PR is ready for review. Feel free to skip python files. These files will get auto-generated in another PRs. For now, a missing parameters were added to have passing all tests and parity between scala and python.

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    "validation_frame",
+    "response_column",
+    "fold_column",
+    "weights_column",
+    "offset_column",
+    "seed",
+    "nfolds",
+    "distribution")
+
+  val unimplemented = Seq(
```
Comment (mn-mikke) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:38]: This list of parameters must be  implement manually. Will do in another PRs. IMHO, this PR is quite big.

File: py/src/ai/h2o/sparkling/ml/algos/H2ODeepLearning.py [hunk 5]
```diff
@@ -22,6 +22,8 @@
 from ai.h2o.sparkling.ml.algos.H2OSupervisedAlgoBase import H2OSupervisedAlgoBase
 from ai.h2o.sparkling.ml.params import H2ODeepLearningParams
 
+import sys
```
Comment (jakubhava) [inline py/src/ai/h2o/sparkling/ml/algos/H2ODeepLearning.py:25]: extra import?

File: ml/src/main/scala/ai/h2o/sparkling/ml/params/H2OAlgoParamsHelper.scala [hunk 6]
```diff
@@ -18,132 +18,6 @@ package ai.h2o.sparkling.ml.params
 
 import java.util
```
Comment (jakubhava) [inline ml/src/main/scala/ai/h2o/sparkling/ml/params/H2OAlgoParamsHelper.scala:19]: nit: maybe rename this class to something like `EnumParamValidator` ?

File: ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OKMeansExtras.scala [hunk 7]
```diff
@@ -16,15 +16,10 @@
...
-import ai.h2o.sparkling.ml.params.H2OKMeansParams
 import ai.h2o.sparkling.{H2OColumnType, H2OFrame}
 import hex.kmeans.KMeansModel.KMeansParameters
-import org.apache.spark.ml.util.{DefaultParamsReadable, Identifiable}
 
-/**
-  * H2O KMeans algorithm exposed via Spark ML pipelines.
-  */
-class H2OKMeans(override val uid: String) extends H2OUnsupervisedAlgorithm[KMeansParameters] with H2OKMeansParams {
+trait H2OKMeansExtras extends H2OAlgorithm[KMeansParameters] {
```
Comment (jakubhava) [inline ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OKMeansExtras.scala:22]: nit: should we limit the scope?

File: ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OTreeBasedSupervisedAlgorithm.scala [hunk 8]
```diff
@@ -18,17 +18,18 @@
...
-abstract class H2OTreeBasedSupervisedAlgorithm[P <: Model.Parameters: ClassTag]
-  extends H2OSupervisedAlgorithm[P]
-  with H2OTreeBasedSupervisedMOJOParams {
+abstract class H2OTreeBasedSupervisedAlgorithm[P <: Model.Parameters: ClassTag] extends H2OSupervisedAlgorithm[P] {
 
   override def fit(dataset: Dataset[_]): H2OTreeBasedSupervisedMOJOModel = {
     super.fit(dataset).asInstanceOf[H2OTreeBasedSupervisedMOJOModel]
   }
+
+  def getNtrees(): Int
```
Comment (jakubhava) [inline ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OTreeBasedSupervisedAlgorithm.scala:32]: Do I understand correctly that the `getNTrees` and the setter are here because of MOJO so they are available there? Otherwise for the wrappers itself they are auto-generated, right?

File: ml/src/main/scala/ai/h2o/sparkling/ml/params/H2OAlgoCommonParams.scala [hunk 9]
```diff
@@ -20,70 +20,54 @@ import ai.h2o.sparkling.ml.params.H2OAlgoParamsHelper.getValidatedEnumValue
...
-      "fold_column" -> getFoldCol(),
-      "keep_cross_validation_predictions" -> getKeepCrossValidationPredictions(),
-      "keep_cross_validation_fold_assignment" -> getKeepCrossValidationFoldAssignment(),
-      "parallelize_cross_validation" -> getParallelizeCrossValidation(),
-      "seed" -> getSeed(),
-      "distribution" -> getDistribution())
+  private[sparkling] override def getH2OAlgorithmParams(): Map[String, Any] = {
+    super.getH2OAlgorithmParams() ++
+      Map(
+        "weights_column" -> getWeightCol(),
```
Comment (jakubhava) [inline ml/src/main/scala/ai/h2o/sparkling/ml/params/H2OAlgoCommonParams.scala:67]: Maybe could we add comment why these can't be auto-generated? EDIT: I see they are mentioned in `IgnoredParams` class, that is fine

Also I see that `weights_column` & `fold_column` have different naming, but couldn't we also auto-generate `seed`, `distribution` & `nfolds` ?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+package ai.h2o.sparkling.api.generation.common
+
+object IgnoredParameters {
+  val deprecated: Seq[String] = Seq(
+    "r2_stopping", // All
+    "max_confusion_matrix_size", // Deep Learning
+    "col_major") // Deep Learning
+
+  val implementedInParent: Seq[String] = Seq(
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:26]: These are the params which have different naming in SW? I was thinking why `seed`, `nfolds` and `distributution` can't be auto-generated

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ai.h2o.sparkling.api.generation.common
+
+object IgnoredParameters {
+  val deprecated: Seq[String] = Seq(
+    "r2_stopping", // All
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:22]: nit: I assume these are deprecated without replacement?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/Parameter.scala [hunk 12]
```diff
@@ -14,19 +14,9 @@
...
-  * H2O GLM algorithm exposed via Spark ML pipelines.
-  */
-class H2OGLM(override val uid: String) extends H2OSupervisedAlgorithm[GLMParameters] with H2OGLMParams {
+case class Parameter(swName: String, h2oName: String, defaultValue: AnyRef, dataType: DataType, comment: String)
 
-  def this() = this(Identifiable.randomUID(classOf[H2OGLM].getSimpleName))
-}
-
-object H2OGLM extends H2OParamsReadable[H2OGLM]
+case class DataType(name: String, fullName: String, isEnum: Boolean)
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/Parameter.scala:22]: is `DataType` Spark Data type?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/Parameter.scala [hunk 13]
```diff
@@ -14,19 +14,9 @@
...
-import ai.h2o.sparkling.ml.utils.H2OParamsReadable
-import hex.glm.GLMModel.GLMParameters
-import org.apache.spark.ml.util.Identifiable
+package ai.h2o.sparkling.api.generation.common
 
-/**
-  * H2O GLM algorithm exposed via Spark ML pipelines.
-  */
-class H2OGLM(override val uid: String) extends H2OSupervisedAlgorithm[GLMParameters] with H2OGLMParams {
+case class Parameter(swName: String, h2oName: String, defaultValue: AnyRef, dataType: DataType, comment: String)
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/Parameter.scala:20]: Do we store the mapping also in generated classes? Especially `swName: String, h2oName: String` will help a lot with GridSearch ( allowing to pass sw param names)

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterNameConverter.scala [hunk 14]
```diff
@@ -14,21 +14,24 @@
...
+package ai.h2o.sparkling.api.generation.common
 
-/**
-  * H2O DeepLearning algorithm exposed via Spark ML pipelines.
-  */
-class H2ODeepLearning(override val uid: String)
-  extends H2OSupervisedAlgorithm[DeepLearningParameters]
-  with H2ODeepLearningParams {
+object ParameterNameConverter {
+  val h2oToSWExceptions: Map[String, String] = Map(
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterNameConverter.scala:21]: I see, exceptions in naming, but still auto-generated

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterNameConverter.scala [hunk 15]
```diff
@@ -14,21 +14,24 @@
...
+    "lambda" -> "lambdaValue",
+    "alpha" -> "alphaValue",
+    "colsample_bylevel" -> "colSampleByLevel",
+    "colsample_bytree" -> "colSampleByTree")
 
-  def this() = this(Identifiable.randomUID(classOf[H2ODeepLearning].getSimpleName))
-}
+  val conversionRules: Map[String, String] = Map("Column" -> "Col")
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterNameConverter.scala:27]: If we have this conversion rule, why `folds_col` and similar params are mentioned also in IgnoredColumns class?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+          s"""  def set${parameter.swName.capitalize}(value: ${resolveParameterType(parameter)}): this.type = {
+             |    set(${parameter.swName}, value)
+             |  }
+           """.stripMargin
+        }
+      }
+      .mkString("\n")
+  }
+
+  private def generateH2OAssignments(parameters: Seq[Parameter]): String = {
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala:131]: Could we include also the opposite mapping in Algorithms (for the grid) so we can pass SW param names?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+import hex.kmeans.KMeansModel.KMeansParameters
+import hex.tree.drf.DRFModel.DRFParameters
+import hex.tree.xgboost.XGBoostModel.XGBoostParameters
+
+object Runner {
+
+  private val explicitDefaultValues = Map("max_w2" -> "java.lang.Float.MAX_VALUE")
+
+  private def parametersConfiguration: Seq[ParameterSubstitutionContext] = {
+    val monotonicity = ExplicitField("monotone_constraints", "HasMonotoneConstraints")
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:38]: Could we do classOf[HasMonotoneConstraints].getSimpleName?

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+import hex.schemas._
+import hex.tree.gbm.GBMModel.GBMParameters
+import hex.glm.GLMModel.GLMParameters
+import hex.kmeans.KMeansModel.KMeansParameters
+import hex.tree.drf.DRFModel.DRFParameters
+import hex.tree.xgboost.XGBoostModel.XGBoostParameters
+
+object Runner {
+
+  private val explicitDefaultValues = Map("max_w2" -> "java.lang.Float.MAX_VALUE")
```
Comment (jakubhava) [inline api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:35]: I'm sure there is a reason for this, I just don't see it :(

File: py/src/ai/h2o/sparkling/ml/algos/H2ODeepLearning.py (hunk 5 above)
Comment (mn-mikke) [reply py/src/ai/h2o/sparkling/ml/algos/H2ODeepLearning.py:25]: it's there because of [the line 119](https://github.com/h2oai/sparkling-water/pull/2065/files/1ae375d10918cbef8b5cf5b39abddb9027da083f#diff-fe929c69a7ddb256b7e016988a4b42d4R119).

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala (hunk 18 above)
Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:35]: When I used the value directly, the scala compiler was complaining that value is too big.

File: py/src/ai/h2o/sparkling/ml/algos/H2ODeepLearning.py (hunk 5 above)
Comment (jakubhava) [reply py/src/ai/h2o/sparkling/ml/algos/H2ODeepLearning.py:25]: I see, thanks!

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala (hunk 17 above)
Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:38]: IMHO, it's not possible. To do that, the `api-generation` project would need to reference the `ml` project. Since `ml` references `api-generation` already. We would get a cyclic reference.

Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:38]: Oki, thanks for explanation!

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala (hunk 18 above)
Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/Runner.scala:35]: ha, interesting behaviour :)

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala (hunk 16 above)
Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala:131]: Np will, do

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/Parameter.scala (hunk 12 above)
Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/Parameter.scala:22]: nevermind, I got it now

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterNameConverter.scala (hunk 15 above)
Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterNameConverter.scala:27]: `folds_col`, `weights_column`, `response_column` and others are ignored since they are already implemented in parent traits (e.g. `H2OAlgoSupervisedParams`, `H2OCommonSupervisedParams`). These traits are also used by grid search and AutoML. Maybe once parameters for Automl and GridSearch will be generated, there will be a room for simplyfing these classes.

Back to your question, this rule is also used for other parameters like `ignored_columns` -> `ignoredCols`

Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/ParameterNameConverter.scala:27]: I see, once grid & automl params are generated as well, we can remove this hiearchy and generate as well. That will be nice next step.

Thanks for explanation!

File: ml/src/main/scala/ai/h2o/sparkling/ml/params/H2OAlgoCommonParams.scala (hunk 9 above)
Comment (mn-mikke) [reply ml/src/main/scala/ai/h2o/sparkling/ml/params/H2OAlgoCommonParams.scala:67]: These parameters are defined in `H2OCommonParams` and utilized also by `GridSearch` and `AutoML`. Once parameters for `GridSearch` and `AutoML` are generated as well, i think we could remove `H2OCommonParams` class and change the hierarchy of traits a bit.

File: ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OTreeBasedSupervisedAlgorithm.scala (hunk 8 above)
Comment (mn-mikke) [reply ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OTreeBasedSupervisedAlgorithm.scala:32]: Yes, you're right NTrees are auto-generated. These methods are here just for ensuring API parity between`H2OTreeBasedSupervisedAlgorithm` and `H2OTreeBasedSupervisedMOJOModel`. You can see usege of these methods in `H2OTreeBasedSupervisedMOJOModelTestSuite.testTrainedTreeBasedModelHasPositiveNumberOfTrees`

File: ml/src/main/scala/ai/h2o/sparkling/ml/params/H2OAlgoParamsHelper.scala (hunk 6 above)
Comment (mn-mikke) [reply ml/src/main/scala/ai/h2o/sparkling/ml/params/H2OAlgoParamsHelper.scala:19]: Np, will rename

File: ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OKMeansExtras.scala (hunk 7 above)
Comment (mn-mikke) [reply ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OKMeansExtras.scala:22]: ok, will do

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala (hunk 10 above)
Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:26]: https://github.com/h2oai/sparkling-water/pull/2065#discussion_r428561558

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala (hunk 11 above)
Comment (mn-mikke) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:22]: The replacement is `stopping_rounds`, `stopping_metric` and `stopping_tolerance`. We removed that manally in `3.30.0.1-1`. https://0xdata.atlassian.net/browse/SW-1819

Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:22]: Thanks!

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala (hunk 10 above)
Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/common/IgnoredParameters.scala:26]: 👍

File: api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala (hunk 16 above)
Comment (jakubhava) [reply api-generation/src/main/scala/ai/h2o/sparkling/api/generation/scala/ParametersTemplate.scala:131]: Thanks! The idea is to have mapping (sw name -> h2o name) so we can translate the sw name to internal h2o name in grid

**Comment by mn-mikke** [comment]
@jakubhava I made same changes to address your comments. Can you have a second look?
